packages/bsky/src/api/app/bsky/unspecced/getPostThreadHiddenV2.ts

```diff
@@ -0,0 +1,118 @@
+import {
+  HydrationFnInput,
+  PresentationFnInput,
+  SkeletonFnInput,
+  createPipeline,
+  noRules,
+} from '../../../../pipeline'
+import { Views } from '../../../../views'
+import { resHeaders } from '../../../util'
+
+const BELOW = 2
+const BRANCHING_FACTOR = 3
+
+export default function (server: Server, ctx: AppContext) {
+  const getPostThreadHidden = createPipeline(
+    skeleton,
+    hydration,
+    noRules, // handled in presentation: 3p block-violating replies are turned to #blockedPost, viewer blocks turned to #notFoundPost.
+    presentation,
+  )
+  server.app.bsky.unspecced.getPostThreadHiddenV2({
+    auth: ctx.authVerifier.optionalStandardOrRole,
+    handler: async ({ params, auth, req }) => {
+      const { viewer, includeTakedowns, include3pBlocks } =
+        ctx.authVerifier.parseCreds(auth)
+      const labelers = ctx.reqLabelers(req)
+      const hydrateCtx = await ctx.hydrator.createContext({
+        labelers,
+        viewer,
+        includeTakedowns,
+        include3pBlocks,
+      })
+
+      return {
+        encoding: 'application/json',
+        body: await getPostThreadHidden({ ...params, hydrateCtx }, ctx),
+        headers: resHeaders({
+          labelers: hydrateCtx.labelers,
+        }),
+      }
+    },
+  })
+}
+
+const skeleton = async (inputs: SkeletonFnInput<Context, Params>) => {
+  const { ctx, params } = inputs
+  const anchor = await ctx.hydrator.resolveUri(params.anchor)
+  try {
+    const res = await ctx.dataplane.getThread({
+      postUri: anchor,
+      above: 0,
+      below: calculateBelow(ctx, anchor, BELOW),
+    })
+    return {
+      anchor,
+      uris: res.uris,
+    }
+  } catch (err) {
+    if (isDataplaneError(err, Code.NotFound)) {
+      return {
+        anchor,
+        uris: [],
+      }
+    } else {
+      throw err
+    }
+  }
+}
+
+const hydration = async (
+  inputs: HydrationFnInput<Context, Params, Skeleton>,
+) => {
+  const { ctx, params, skeleton } = inputs
+  return ctx.hydrator.hydrateThreadPosts(
+    skeleton.uris.map((uri) => ({ uri })),
+    params.hydrateCtx,
+  )
+}
+
+const presentation = (
+  inputs: PresentationFnInput<Context, Params, Skeleton>,
+) => {
+  const { ctx, params, skeleton, hydration } = inputs
+  const thread = ctx.views.threadHiddenV2(skeleton, hydration, {
+    below: calculateBelow(ctx, skeleton.anchor, BELOW),
+    branchingFactor: BRANCHING_FACTOR,
+    viewer: params.hydrateCtx.viewer,
+  })
+  return { thread }
+}
+
+type Context = {
+  dataplane: DataPlaneClient
+  hydrator: Hydrator
+  views: Views
+  cfg: ServerConfig
+}
+
+type Params = QueryParams & { hydrateCtx: HydrateCtx }
+
+type Skeleton = {
+  anchor: string
+  uris: string[]
+}
+
+const calculateBelow = (ctx: Context, anchor: string, below: number) => {
+  let maxDepth = ctx.cfg.maxThreadDepth
+  if (ctx.cfg.bigThreadUris.has(anchor) && ctx.cfg.bigThreadDepth) {
+    maxDepth = ctx.cfg.bigThreadDepth
+  }
+  return maxDepth ? Math.min(maxDepth, below) : below
+}
```

Comment:
I think this is shared by 2-3 endpoints now.  Could be nice to factor out somewhere.